### ISSUE TITLE
ncurses and netcdf fails to build with syntax error on Cray

### ISSUE THREAD

**Issue by adamryczkowski** [spack/spack#6861]
### Summary

ncurses fails to build with syntax error on Cray

### Actual Result

```
$ spack install ncurses
==> pkgconf is already installed in /users/adamr/tmp/spack/opt/spack/cray-CNL-haswell/gcc-7.1.0/pkgconf-1.3.10-uhipjojjl7jb2zoq5e3lbcd6fhmelmse
==> Installing ncurses
==> Error: SyntaxError: invalid syntax (<string>, line 476)
SyntaxError: SyntaxError: invalid syntax (<string>, line 476)

/users/adamr/tmp/spack/lib/spack/spack/build_environment.py:683, in child_process:
     27    
     28                # build up some context from the offending package so we can
     29                # show that, too.
  >> 30                package_context = get_package_context(tb)
     31    
     32                build_log = None
     33                if hasattr(pkg, 'log_path'):
```

```
$ spack install netcdf %gcc@7.1.0
==> libsigsegv is already installed in /users/adamr/tmp/spack/opt/spack/cray-CNL-haswell/gcc-7.1.0/libsigsegv-2.11-tatqezxww74hs74fsldkwvpy6kf5isjp
==> m4 is already installed in /users/adamr/tmp/spack/opt/spack/cray-CNL-haswell/gcc-7.1.0/m4-1.4.18-uuicyh5c6ms2gfbfklywbkrqujjvepbe
==> libtool is already installed in /users/adamr/tmp/spack/opt/spack/cray-CNL-haswell/gcc-7.1.0/libtool-2.4.6-sy3pq3d4f466lgiggepp3meln2tsomrx
==> pkgconf is already installed in /users/adamr/tmp/spack/opt/spack/cray-CNL-haswell/gcc-7.1.0/pkgconf-1.3.10-uhipjojjl7jb2zoq5e3lbcd6fhmelmse
==> util-macros is already installed in /users/adamr/tmp/spack/opt/spack/cray-CNL-haswell/gcc-7.1.0/util-macros-1.19.1-exb3t3uusocfq3ou63vbscihaf2sq5i5
==> Installing libpciaccess
==> Error: SyntaxError: invalid syntax (<string>, line 476)
SyntaxError: SyntaxError: invalid syntax (<string>, line 476)

/users/adamr/tmp/spack/lib/spack/spack/build_environment.py:683, in child_process:
     27    
     28                # build up some context from the offending package so we can
     29                # show that, too.
  >> 30                package_context = get_package_context(tb)
     31    
     32                build_log = None
     33                if hasattr(pkg, 'log_path'):
```

### Steps to reproduce the issue

Well, I wanted to build the `meson` package, since it is missing on my data center. I pulled the recent Spack, and typed `spack install meson`. `ncurses` happen to be on of the first dependencies for the meson.

The problem persists on `gcc@4.9.3`, `gcc@7.1.0` and `gcc@5.3.0`.

The specs:

```
$ uname -r
4.4.74-92.38-default

$ lsb_release -a
LSB Version:	n/a
Distributor ID:	SUSE
Description:	SUSE Linux Enterprise Server 12 SP2
Release:	12.2
Codename:	n/a

$ python --version
Python 2.7.13
```

**Comment by adamjstewart** [comment]
I'm confused by the `SyntaxError`. This suggests that there was a problem with Spack, not with the package you were trying to install. Can you run with `spack --debug --verbose` and see if it gives you a more specific error message?

**Comment by adamryczkowski** [comment]
This is the output of the command `spack --debug --verbose install ncurses`:

```
==> Reading config file /users/adamr/tmp/spack/etc/spack/defaults/packages.yaml
==> READ LOCK: /users/adamr/.spack/cache/providers/.builtin-index.yaml.lock[0:0] [Acquiring]
==> READ LOCK: /users/adamr/.spack/cache/providers/.builtin-index.yaml.lock[0:0] [Acquired]
==> READ LOCK: /users/adamr/.spack/cache/providers/.builtin-index.yaml.lock[0:0] [Released]
==> Reading config file /users/adamr/.spack/cray/compilers.yaml
==> WRITE LOCK: /users/adamr/tmp/spack/opt/spack/.spack-db/prefix_lock[1758063763337378993:1] [Acquiring]
==> WRITE LOCK: /users/adamr/tmp/spack/opt/spack/.spack-db/prefix_lock[1758063763337378993:1] [Acquired]
==> READ LOCK: /users/adamr/tmp/spack/opt/spack/.spack-db/lock[0:0] [Acquiring]
==> READ LOCK: /users/adamr/tmp/spack/opt/spack/.spack-db/lock[0:0] [Acquired]
==> READ LOCK: /users/adamr/tmp/spack/opt/spack/.spack-db/lock[0:0] [Released]
==> WRITE LOCK: /users/adamr/tmp/spack/opt/spack/.spack-db/prefix_lock[1758063763337378993:1] [Released]
==> link /users/adamr/tmp/spack/var/spack/stage/ncurses-6.0-gdf4wwzunp4whfm62yktzu4w7ejabt3a -> /tmp/adamr/spack-stage/spack-stage-H0tuVV
==> READ LOCK: /users/adamr/tmp/spack/opt/spack/.spack-db/prefix_lock[1758063763337378993:1] [Acquiring]
==> READ LOCK: /users/adamr/tmp/spack/opt/spack/.spack-db/prefix_lock[1758063763337378993:1] [Acquired]
==> READ LOCK: /users/adamr/tmp/spack/opt/spack/.spack-db/prefix_lock[1758063763337378993:1] [Released]
==> Installing ncurses dependencies
==> WRITE LOCK: /users/adamr/tmp/spack/opt/spack/.spack-db/prefix_lock[5830044255681243406:1] [Acquiring]
==> WRITE LOCK: /users/adamr/tmp/spack/opt/spack/.spack-db/prefix_lock[5830044255681243406:1] [Acquired]
==> WRITE LOCK: /users/adamr/tmp/spack/opt/spack/.spack-db/prefix_lock[5830044255681243406:1] [Released]
==> link /users/adamr/tmp/spack/var/spack/stage/pkgconf-1.3.10-uhipjojjl7jb2zoq5e3lbcd6fhmelmse -> /tmp/adamr/spack-stage/spack-stage-lBF_im
==> READ LOCK: /users/adamr/tmp/spack/opt/spack/.spack-db/prefix_lock[5830044255681243406:1] [Acquiring]
==> READ LOCK: /users/adamr/tmp/spack/opt/spack/.spack-db/prefix_lock[5830044255681243406:1] [Acquired]
==> pkgconf is already installed in /users/adamr/tmp/spack/opt/spack/cray-CNL-haswell/gcc-7.1.0/pkgconf-1.3.10-uhipjojjl7jb2zoq5e3lbcd6fhmelmse
==> READ LOCK: /users/adamr/tmp/spack/opt/spack/.spack-db/prefix_lock[5830044255681243406:1] [Released]
==> Installing ncurses
==> '/opt/cray/pe/modules/3.2.10.6/bin/modulecmd' 'python' 'list'
==> '/opt/cray/pe/modules/3.2.10.6/bin/modulecmd' 'python' 'avail' '-t' 'craype-'
==> '/usr/bin/env' '-' 'USER=adamr' 'HOME=/users/adamr' '/bin/bash' '--noprofile' '--norc' '-c' '. /etc/profile; module list -lt'
==> Found default modules:
       modules/3.2.10.6
       eproxy/2.0.16-6.0.4.1_3.1__g001b199.ari
       cce/8.6.1
       craype-network-aries
       craype/2.5.12
       cray-libsci/17.06.1
       udreg/2.3.2-6.0.4.0_12.2__g2f9c3ee.ari
       ugni/6.0.14-6.0.4.0_14.1__ge7db4a2.ari
       pmi/5.0.12
       dmapp/7.1.1-6.0.4.0_46.2__gb8abda2.ari
       gni-headers/5.0.11-6.0.4.0_7.2__g7136988.ari
       xpmem/2.2.2-6.0.4.0_3.1__g43b0535.ari
       job/2.2.2-6.0.4.0_8.2__g3c644b5.ari
       dvs/2.7_2.2.32-6.0.4.1_7.1__ged1923a
       alps/6.4.1-6.0.4.0_7.2__g86d0f3d.ari
       rca/2.2.11-6.0.4.0_13.2__g84de67a.ari
       atp/2.1.1
       perftools-base/6.5.1
       PrgEnv-cray/6.0.4
       cray-mpich/7.6.0
       slurm/17.02.9-1
       craype-haswell
       xalt/daint-2016.11
       libffi/.3.2.1
       ncurses/.6.0
       gettext/.0.19.8
       zlib/.1.2.8
       XZ/.5.2.2
       libxml2/.2.9.4
       PCRE/.8.38
       GLib/.2.49.5
       S-Lang/2.3.1a
       mc/4.8.18
       libevent/.2.0.22
       libutempter/.1.1.6-3
       tmux/2.3
==> '/opt/cray/pe/modules/3.2.10.6/bin/modulecmd' 'python' 'list'
==> '/opt/cray/pe/modules/3.2.10.6/bin/modulecmd' 'python' 'show' 'gcc/7.1.0'
==> '/opt/cray/pe/modules/3.2.10.6/bin/modulecmd' 'python' 'list'
==> '/opt/cray/pe/modules/3.2.10.6/bin/modulecmd' 'python' 'show' 'gcc/7.1.0'
==> '/opt/cray/pe/modules/3.2.10.6/bin/modulecmd' 'python' 'list'
==> '/opt/cray/pe/modules/3.2.10.6/bin/modulecmd' 'python' 'show' 'PrgEnv-gnu'
==> '/opt/cray/pe/modules/3.2.10.6/bin/modulecmd' 'python' 'unload' 'PrgEnv'
==> '/opt/cray/pe/modules/3.2.10.6/bin/modulecmd' 'python' 'unload' 'PrgEnv-x1'
==> '/opt/cray/pe/modules/3.2.10.6/bin/modulecmd' 'python' 'unload' 'PrgEnv-x2'
==> '/opt/cray/pe/modules/3.2.10.6/bin/modulecmd' 'python' 'unload' 'PrgEnv-cray'
==> Error: SyntaxError: invalid syntax (<string>, line 476)
SyntaxError: SyntaxError: invalid syntax (<string>, line 476)

/users/adamr/tmp/spack/lib/spack/spack/build_environment.py:683, in child_process:
     27    
     28                # build up some context from the offending package so we can
     29                # show that, too.
  >> 30                package_context = get_package_context(tb)
     31    
     32                build_log = None
     33                if hasattr(pkg, 'log_path'):


Traceback (most recent call last):
  File "/users/adamr/tmp/spack/lib/spack/spack/build_environment.py", line 664, in child_process
    setup_package(pkg, dirty=dirty)
  File "/users/adamr/tmp/spack/lib/spack/spack/build_environment.py", line 616, in setup_package
    load_module(mod)
  File "/users/adamr/tmp/spack/lib/spack/spack/util/module_cmd.py", line 134, in load_module
    error=str), '<string>', 'exec'))
  File "<string>", line 476
    rca/2.2.11-6.0.4.0_13.2__g84de67a.ari(21):ERROR:102: Tcl command execution failed: if { [ file exists /opt/cray/rca/2.2.11-6.0.4.0_13.2__g84de67a.ari/lib64/pkgconfig/cray-rca.pc ] } {
             ^
SyntaxError: invalid syntax
```

**Comment by adamjstewart** [comment]
It looks like this is a problem with our module support on Cray. Unfortunately, I don't have any experience working on Cray. Pinging @alalazo @pramodk @pramodskumbhar @skosukhin @becker33 @scheibelp as one of them should be able to help.

**Comment by mamelara** [comment]
What system are you on? I'm on Cori and am not encountering this issue.

**Comment by adamryczkowski** [comment]
After the Spack update the error still persists. Here are the specs of the supercomputer: https://www.cscs.ch/computers/piz-daint

**Comment by mamelara** [comment]
Got it. Was able to replicate. So I remember we had issues with pkg-config working on Cray and we added some setup to the environment to make it work (spack edit pkg-config) It looks like pkgconf might not have the right setup for Cray.

I recommend either declaring pkg-config in your packages.yaml and using the external one or specifying a preference to pkg-config over pkgconf to satisfy the virtual dependency to pkgconfig as a quick fix. I've never used pkgconf so not sure if the same solution we did for pkg-config would work the same.

```yaml
packages:
    all:
        providers:
            pkgconfig: [pkg-config]
```
